src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -611,15 +558,49 @@ private static string UrlPathEncodeImpl(string value)
                 return value;
             }
 
-            // recurse in case there is a query string
-            int i = value.IndexOf('?');
-            if (i >= 0)
+            int i = value.AsSpan().IndexOfAnyExceptInRange((char)0x21, (char)0x7F);
+            if (i < 0)
+            {
+                return value;
+            }
+
+            int indexOfQuery = value.IndexOf('?');
+            if ((uint)indexOfQuery < (uint)i)
             {
-                return string.Concat(UrlPathEncodeImpl(value.Substring(0, i)), value.AsSpan(i));
+                // Everything before the Query is valid ASCII
+                return value;
+            }
+
+            ReadOnlySpan<char> toEncode = indexOfQuery >= 0
+                ? value.AsSpan(i, indexOfQuery - i)
+                : value.AsSpan(i);
+
+            byte[] bytes = ArrayPool<byte>.Shared.Rent(Encoding.UTF8.GetMaxByteCount(toEncode.Length));
+            char[] chars = ArrayPool<char>.Shared.Rent(bytes.Length * 3);
+
+            int utf8Length = Encoding.UTF8.GetBytes(toEncode, bytes);
```

Comment:
```suggestion
            int utf8Length = Encoding.UTF8.GetBytes(toEncode, bytes);
            char[] chars = ArrayPool<char>.Shared.Rent(utf8Length * 3);
```
Nit: we can swap the two to potentially rent a smaller buffer